Cython/Utility/ObjectHandling.c

```diff
@@ -2217,7 +2217,16 @@ static PyObject* __Pyx_PyObject_CallMethod1(PyObject* obj, PyObject* method_name
 
 #define __Pyx_tp_new(type_obj, args) __Pyx_tp_new_kwargs(type_obj, args, NULL)
 static CYTHON_INLINE PyObject* __Pyx_tp_new_kwargs(PyObject* type_obj, PyObject* args, PyObject* kwargs) {
+#if CYTHON_COMPILING_IN_LIMITED_API
+    PyObject *tp_new, *obj;
+    tp_new = PyObject_GetAttrString(type_obj, "__new__");
+    if (unlikely(!tp_new)) return NULL;
+    obj = PyObject_Call(tp_new, args, kwargs);
+    Py_DECREF(tp_new);
+    return obj;
+#else
     return (PyObject*) (((PyTypeObject*)type_obj)->tp_new((PyTypeObject*)type_obj, args, kwargs));
+#endif
```

Comment:
In principle I think we can do better on Limited API >= 3.10:

```
#if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030A0000
    // ... your code
#else
    return (PyObject*) __Pyx_PyType_GetSlot((PyTypeObject*)type_obj, tp_new, newfunc)((PyTypeObject*)type_obj, args, kwargs);
#endif
```

Reply:
Or

```
PyObject *result = (PyObject*) __Pyx_PyType_TryGetSlot((PyTypeObject*)type_obj, tp_new, newfunc)((PyTypeObject*)type_obj, args, kwargs);
#if (CYTHON_LIMITED_API &&  __PYX_LIMITED_VERSION_HEX < 0x030A0000
if (!result) {
  // Your code
}
#endif
return result;
```

which will be faster on heap types on all limited API versions

Reply:
Oh, that's great! Thanks for the pointer! Actually, what I've done so far does not actually work, getattr "__new__" returns a wrapper callable that in the end does not behave as expected and raises `TypeError`.